docs/porting-example/steps/step_02_hpy_legacy.c

```diff
@@ -0,0 +1,173 @@
+// have been converted to HPy methods that receive handles as arguments, but
+// other methods are still legacy methods that receive PyObject arguments.
+
+typedef struct {
+    // PyObject_HEAD is required while legacy methods still access
+    // PointObject and should be removed once the port to HPy is completed.
+    PyObject_HEAD
+    double x;
+    double y;
+    // HPy handles are shortlived to support all GC strategies
+    // For that reason, PyObject* in C structs are replaced by HPyField
+    HPyField obj;
+} PointObject;
+
+// This defines PyPointObject as an alias of PointObject so that existing
+// code that still uses PyPointObject and expects PyObject_HEAD continues to
+// compile and run. Once PyObject_HEAD has been removed, this alias should be
+// removed so that code that still expects PyObject_HEAD will fail to compile.
+typedef PointObject PyPointObject;
+
+// The legacy type helper macro defines an PointObject_AsStruct function allows
+// non-legacy methods to convert HPy handles to PointObject structs. The legacy
+// type helper macro is used because PyObject_HEAD is still present in
+// PointObject. Once PyObject_HEAD has been removed (see point_hpy_final.c) we
+// will use HPy_TYPE_HELPERS instead.
+HPyType_LEGACY_HELPERS(PointObject)
+
+HPyDef_SLOT(Point_traverse, Point_traverse_impl, HPy_tp_traverse)
+int Point_traverse_impl(void *self, HPyFunc_visitproc visit, void *arg)
+{
+    HPy_VISIT(&((PointObject*)self)->obj);
+    return 0;
+}
+
+// this is a method for creating a Point
+HPyDef_SLOT(Point_init, Point_init_impl, HPy_tp_init)
+int Point_init_impl(HPyContext *ctx, HPy self, HPy *args, HPy_ssize_t nargs, HPy kw)
+{
+    static const char *kwlist[] = {"x", "y", "obj", NULL};
+    PointObject *p = PointObject_AsStruct(ctx, self);
+    p->x = 0.0;
+    p->y = 0.0;
+    HPy obj = HPy_NULL;
+    HPyTracker ht;
+    if (!HPyArg_ParseKeywords(ctx, &ht, args, nargs, kw, "|ddO", kwlist,
+                              &p->x, &p->y, &obj))
+        return -1;
+    if (HPy_IsNull(obj))
+        obj = ctx->h_None;
+    // INCREF not needed because HPyArg_ParseKeywords does not steal a reference
+    HPyField_Store(ctx, self, &p->obj, obj);
+    HPyTracker_Close(ctx, ht);
+    return 0;
+}
+
+// this is the getter for the associated object
+HPyDef_GET(Point_obj_get, "obj", Point_obj_get_impl, .doc="Associated object.")
+HPy Point_obj_get_impl(HPyContext *ctx, HPy self, void* closure)
+{
+    PointObject *p = PointObject_AsStruct(ctx, self);
+    return HPyField_Load(ctx, self, p->obj);
+}
+
+// an HPy method of Point
+HPyDef_METH(Point_norm, "norm", Point_norm_impl, HPyFunc_NOARGS, .doc="Distance from origin.")
+HPy Point_norm_impl(HPyContext *ctx, HPy self)
+{
+    PointObject *p = PointObject_AsStruct(ctx, self);
+    double norm;
+    HPy result;
+    norm = sqrt(p->x * p->x + p->y * p->y);
+    result = HPyFloat_FromDouble(ctx, norm);
+    return result;
```

Comment:
This could just be `return HPyFloat_FromDouble(ctx, norm);`. Variable `HPy result` is not required.